srk/src/terminationLLRF.ml

```diff
@@ -0,0 +1,194 @@
+    let poly = Polyhedron.of_formula x_space_cs formula_of_lbx in
+    let constraints = Polyhedron.enum_constraints poly in
+    let lb_cone_generators = BatEnum.map (
+      fun (constraint_kind, v) -> 
+        match constraint_kind with 
+        | `Nonneg | `Pos -> let _, w = Linear.QQVector.pivot (-1) v in (`Ray, w)
+        | `Zero -> let _, w = Linear.QQVector.pivot (-1) v in (`Line, w)
+      ) 
+      constraints 
+    in
+    (* Origin must belong to the lower-bounded terms cone *)
+    BatEnum.push lb_cone_generators (`Vertex, Linear.QQVector.zero);
+    let generators_list = BatList.of_enum lb_cone_generators in
+    let lb_cone = Polyhedron.of_generators dual_space_dim (BatList.enum generators_list) in
+    lb_cone
+  in
+
+  logf "calculating meet of non-inc cone and lower-bounded cone";
+  let non_inc_cone = compute_non_inc_cone formula in
+  if Polyhedron.equal non_inc_cone Polyhedron.bottom then
+  begin
+    logf ~attributes:[`Bold; `Red] "Non-increasing term cone is empty, LLRF fail";
+    (false, formula) 
+  end
+  else 
+  begin
+    let lb_cone = compute_lower_bounded_cone formula in
+    if Polyhedron.equal non_inc_cone Polyhedron.bottom then
+      begin
+        logf ~attributes:[`Bold; `Red] "Lower-bounded term cone is empty, LLRF fail";
+        (false, formula) 
+      end
+    else 
+      begin
+        let quasi_rf_cone = Polyhedron.meet non_inc_cone lb_cone in
+        logf "qrf cone: %a" (Formula.pp srk) (Polyhedron.to_formula dual_space_cs quasi_rf_cone);
+        if Polyhedron.equal quasi_rf_cone Polyhedron.bottom then
+          begin
+            logf ~attributes:[`Bold; `Red] "Quasi-ranking function cone is empty, LLRF fail";
+            (false, formula) 
+          end
+        else 
+          begin 
+            let qrf_cone_generators = BatList.of_enum (Polyhedron.enum_generators dual_space_dim quasi_rf_cone) in
+            let coeff_all_zero = not (BatList.exists (fun (typ, _) -> match typ with | `Ray | `Line -> true | _ -> false) qrf_cone_generators) in
+            if coeff_all_zero then 
+              begin
+                logf ~attributes:[`Bold; `Red] "only all zero quasi ranking function exists at this level, fail";
+                (false, formula) 
+              end
+            else
+              begin
+                let qrf_unchange_constraints = BatList.map 
+                  (fun (gen_kind, vec) -> 
+                    match gen_kind with 
+                    | `Vertex -> mk_true srk
+                    | `Ray ->
+                      logf "looking at generator with vector: %a" Linear.QQVector.pp vec;
+                      let unprimed_exp = Linear.term_of_vec srk (fun d -> let x, _ = BatList.nth x_xp d in mk_const srk x) vec in
+                      let primed_exp = Linear.term_of_vec srk (fun d -> let _, xp = BatList.nth x_xp d in mk_const srk xp) vec in
+                      mk_eq srk unprimed_exp primed_exp
+                    | `Line -> 
```

Comment:
A line corresponds to a constant term (both vec and -vec are non-increasing, and therefore vec does not change).  Why do we equate vec' with 0?